packages/template-retail-react-app/app/utils/product-utils.js

```diff
@@ -188,20 +163,56 @@ export const getDecoratedVariationAttributes = (product, opts = {}) => {
 }
 
 /**
-     
+ * Find the lowest price of a product, across all of its price books and promotions
+ * @param {Object} product - product data from the API
+ * @returns {Object|undefined} the lowest price for the given product. If it's a promotional price, the promotion will also be returned.
+ */
+export const findLowestPrice = (product) => {
+    if (!product) return
+    const array = product.variants ?? [product]
+
+    return array.reduce(
+        (prev, data) => {
+            const promotions = data.productPromotions || []
+            const [smallestPromotionalPrice, promo] = getSmallestValByProperty(
+                promotions,
+                'promotionalPrice'
+            )
+
+            let salePrice = data.price
+            let promotion = null
+            if (smallestPromotionalPrice && smallestPromotionalPrice < data.price) {
+                salePrice = smallestPromotionalPrice
+                promotion = promo
+            }
+
+            return salePrice < prev.minPrice ? {minPrice: salePrice, promotion, data} : prev
+        },
+        {minPrice: Infinity, promotion: null, data: null}
+    )
+}
+
+/**
  * @private
  * Find the smallest value by key from a given array
- * @param arr
- * @param key
+ * @returns {Array} an array of such smallest value and the item containing this value
+ * @example
+ * const [value, itemContainingValue] = getSmallestValByProperty(array, key)
  */
 const getSmallestValByProperty = (arr, key) => {
-    if (!arr || !arr.length) return undefined
+    if (!arr || !arr.length) return []
     if (!key) {
         throw new Error('Please specify a key.')
     }
-    const vals = arr
-        .map((item) => item[key])
-        .filter(Boolean)
-        .filter(Number)
-    return vals.length ? Math.min(...vals) : undefined
+    const filtered = arr.filter((item) => Boolean(item[key])).filter((item) => Number(item[key]))
```

Comment:
Nit: shouldn't filter by Number() will rule out falsy value too? 🤔

Reply:
In this case, I was following what's been done before: first filter with Boolean and then by Number.

Reply:
ah gotcha, okay confirming that we don't need to filter by Boolean here since filter Number will already rule out falsy values. Checked in browser console. 
```
const test = [null, undefined, 1, 2, 5]
test.filter(Number)
=> [1, 2, 5]
```

Reply:
@alexvuong thanks for confirming it. There's something else, however. It was possible to ignore 0 in the calculation.

Now in the latest commit, if 0 is one of the values, it will be included.